privacyidea/lib/utils.py

```diff
@@ -1192,4 +1192,56 @@ def split_pin_pass(passw, otplen, prependpin):
         otpval = passw[0:otplen]
         log.debug("PIN appended. PIN length is {0!s}, OTP length is {0!s}.".format(len(pin),
                                                                                    len(otpval)))
-    return pin, otpval
\ No newline at end of file
+    return pin, otpval
+
+
+def create_tag_dict(logged_in_user=None,
+                    request=None,
+                    serial=None,
+                    tokenowner=None,
+                    tokentype=None,
+                    recipient=None,
+                    registrationcode=None,
+                    googleurl_value=None,
+                    client_ip=None):
+    """
+    This helper function creates a dictionary with tags to be used in sending emails
+    either with email tokens or within the notification handler
+
+    :param logged_in_user: The acting logged in user (admin)
+    :param request: The HTTP request object
+    :param serial: The serial number of the token
+    :param tokenowner: The owner of the token
+    :type tokenowner: user object
+    :param tokentype: The type of the token
+    :param recipient: The recipient
+    :type recipient: dictionary with "givenname" and "surname"
+    :param registrationcode: The registration code of a token
+    :param googleurl_value: The URL for the QR code during token enrollemnt
+    :param client_ip: The IP of the client
+    :return: The tag dictionary
+    """
+    time = datetime.now().strftime("%H:%M:%S")
+    date = datetime.now().strftime("%Y-%m-%d")
+    recipient = recipient or {}
+    tags = dict(admin=logged_in_user.get("username") if logged_in_user else "",
+                realm=logged_in_user.get("realm") if logged_in_user else "",
+                action=request.path if request else "",
+                serial=serial,
+                url=request.url_root if request else "",
+                user=tokenowner.info.get("givenname") if tokenowner else "",
+                surname=tokenowner.info.get("surname") if tokenowner else "",
+                givenname=recipient.get("givenname"),
+                username=tokenowner.login,
+                userrealm=tokenowner.realm,
+                tokentype=tokentype,
+                registrationcode=registrationcode,
+                recipient_givenname=recipient.get("givenname"),
+                recipient_surname=recipient.get("surname"),
+                googleurl_value=googleurl_value,
+                time=time,
+                date=date,
+                client_ip=client_ip,
+                ua_browser=request.user_agent.browser if request else "",
+                ua_string=request.user_agent.string if request else "")
```

Comment:
I think defining a ``create_tag_dict`` function is a good idea!
But I was thinking: The Email token and the user notification handler are able to send HTML mails, and here we fill the HTML template with user-controlled data, e.g. ``ua_string``. So we should probably escape all HTML tags in the user-controlled dictionary values, e.g. ``ua_string``, ``ua_browser``, ``url``, ``user``, ``surname`` ... 
What do you think?

Reply:
The surname is not user controlled. It is admin controlled. 
I think the username "user" is also not user controlled.
I think it is only the componentes of ``request``, right?

Do we have a method for this?

Reply:
We can use [flask.escape](http://flask.pocoo.org/docs/1.0/api/#flask.escape) or [cgi.escape](https://docs.python.org/2/library/cgi.html#cgi.escape), if we want to avoid the flask import.

Yes, we should definitely escape the components of ``request``.

But can we be sure that the surname, user, ... are always trusted, or may there be cases in which users can edit their own names? I was thinking if we should just escape *all* values, just to be sure:
```python
escaped_tags = {}
for key, value in tags:
    escaped_tags[key] = cgi.escape(value) if value is not None else value
```
What do you think?

Reply:
Hm I do not like being address in the email as "Cornelius K\xf6lbel".